File: web-cardr/src/lib/engagement.ts
import { differenceInDays } from "date-fns";

export type EngagementTier = "A" | "B" | "C";

export interface EngagementScore {
  tier: EngagementTier;
  label: string;
  color: string;
  bgColor: string;
  daysSinceActivity: number | null;
}

/**
 * Calculate engagement tier based on most recent activity across:
 * - next_action_date (upcoming = active)
 * - follow_up_date
 * - last activity timestamp
 * - scanned_at (fallback)
 *
 * A = Hot (activity within 7 days)
 * B = Warm (8–30 days)
 * C = Cold (31+ days or no activity)
 */
export function getEngagementScore(contact: {
  scannedAt?: string;
  followUpDate?: string;
  nextActionDate?: string;
  enrichedAt?: string;
}, lastActivityDate?: string | null): EngagementScore {
  const now = new Date();
  const dates: Date[] = [];

  // Collect all relevant dates
  if (lastActivityDate) dates.push(new Date(lastActivityDate));
  if (contact.enrichedAt) dates.push(new Date(contact.enrichedAt));
  if (contact.scannedAt) dates.push(new Date(contact.scannedAt));

  // Upcoming next action date counts as very active
  if (contact.nextActionDate) {
    const nad = new Date(contact.nextActionDate);
    if (nad >= now) {
      // Future action date = hot
      return { tier: "A", label: "Hot", color: "text-emerald-400", bgColor: "bg-emerald-400/10", daysSinceActivity: 0 };
    }
    dates.push(nad);
  }

  if (contact.followUpDate) {
    const fd = new Date(contact.followUpDate);
    if (fd >= now) {
      return { tier: "A", label: "Hot", color: "text-emerald-400", bgColor: "bg-emerald-400/10", daysSinceActivity: 0 };
    }
    dates.push(fd);
  }

  if (dates.length === 0) {
    return { tier: "C", label: "Cold", color: "text-zinc-400", bgColor: "bg-zinc-400/10", daysSinceActivity: null };
  }

  const mostRecent = new Date(Math.max(...dates.map((d) => d.getTime())));
  const days = differenceInDays(now, mostRecent);

  if (days <= 7) {
    return { tier: "A", label: "Hot", color: "text-emerald-400", bgColor: "bg-emerald-400/10", daysSinceActivity: days };
  }
  if (days <= 30) {
    return { tier: "B", label: "Warm", color: "text-amber-400", bgColor: "bg-amber-400/10", daysSinceActivity: days };
  }
  return { tier: "C", label: "Cold", color: "text-zinc-400", bgColor: "bg-zinc-400/10", daysSinceActivity: days };
}

File: ios-cardr/CardrAI/Utilities/Engagement.swift
import SwiftUI

/// Engagement tier for a contact, mirroring the web `getEngagementScore`.
/// A = Hot (activity within 7 days), B = Warm (8–30 days), C = Cold (31+ / none).
enum EngagementTier: String {
    case a = "A"
    case b = "B"
    case c = "C"

    var label: String {
        switch self {
        case .a: "Hot"
        case .b: "Warm"
        case .c: "Cold"
        }
    }

    var color: Color {
        switch self {
        case .a: Theme.success
        case .b: Theme.warning
        case .c: Color(hex: "8A8A93")
        }
    }
}

enum Engagement {
    private static let parser = ISO8601DateFormatter()

    private static func date(_ raw: String?) -> Date? {
        guard let raw else { return nil }
        if let d = parser.date(from: raw) { return d }
        // Fallback for plain dates without time component.
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f.date(from: String(raw.prefix(10)))
    }

    /// Scores a contact based on its most recent relevant date. Mirrors the web logic
    /// using the fields available natively (follow-up, scanned, created).
    static func tier(for contact: Contact) -> EngagementTier {
        let now = Date()

        // Future follow-up = actively engaged.
        if let follow = date(contact.followUpDate), follow >= now {
            return .a
        }

        var dates: [Date] = []
        if let follow = date(contact.followUpDate) { dates.append(follow) }
        if let scanned = date(contact.scannedAt) { dates.append(scanned) }
        if let created = date(contact.createdAt) { dates.append(created) }

        guard let mostRecent = dates.max() else { return .c }
        let days = Calendar.current.dateComponents([.day], from: mostRecent, to: now).day ?? 999
        if days <= 7 { return .a }
        if days <= 30 { return .b }
        return .c
    }
}
